### ISSUE TITLE
Add helper script for generating attributes file boilerplate from config files

### ISSUE THREAD

**Issue by patcon** [sczizzo/magic-cookbook#2]
I've been meaning to create this for awhile, but never quite got motivated. Perhaps you might be a good person to offer feedback, as it seems like it complements this cookbooks goals.

This stems from the decision that I liked generating yaml/json/ini config dynamically via attributes namespace (ie. `node["mycookbook"]["config"]["key"] = "value"`) rather than using included erb templates. I assume the maintainer of this cookbook probably agrees.

The thought was that it's tedious to generate the attributes.rb file the first time, and downright inconvenient for users trying to migrate old config files (or config files from tutorials) into a format this sort of cookbook accommodates. Also, kinda tedious to understand updates to the default config when, for example, migrating between major version in the package repo, as diffs are no longer useful when the format has been completely changed to attributes.rb

Anyhow, think a little helper ruby script (for running outside the chef context) would make sense to include. Would likely fit as a Thorfile or Rakefile, as I imagine it.

Appreciate any feedback on whether you'd find it useful

**Comment by sczizzo** [comment]
I agree, it's certainly inconvenient and tedious to migrate. We made the transition to this style by adding shims in our ERB templates, so we could use the new "attribute-driven" style alongside the old "hardcoded" stuff:
```
# attributes/default.rb
default['config'] = {
  'abc' => 'foo',
  'xyz' => 'bar'
}
default['config_shim'] = {
  'ns3' => {
    'baz' => 'bap'
  }
}
```
```
# templates/default/example.ini.erb
[ns1]
abc=<%= node['config']['abc'] %>

[ns2]
xyz=<%= node['config']['xyz'] %>

<% node['config_shim'].each do |namespace, config| %>
[<%= namespace %>]
<% config.each do |key, value| %>
<%= key %>=<%= value %>
<% end %>
<% end %>
```
Painful, but we were rewriting all our Chef code anyway.

I will say now that we have this kind of general, autogenerated configuration, we're able to build tools on top so we don't necessarily define attributes in Chef at all. So I think it's a useful concept, and the code I've written for the config generators here in the Magic cookbook are just plain Ruby. They could easily be abstracted out into a library gem for import in a Rakefile or any Ruby project.

We've also been thinking about how to manage configuration changes over time, and I've got a little experiment around defining "attribute schemas" and then creating a conformant document in JSON: https://github.com/sczizzo/smort (sorry it's very rough, I really just started this past week). I have some ideas around versioning the schemas, so Chef will raise an exception if you try to deploy an application with a configuration scheme it doesn't support

tl;dr Yeah, totally agree

**Comment by patcon** [comment]
Thanks!

> we're able to build tools on top so we don't necessarily define attributes in Chef at all

At risk of side-tracking, can I ask what you're using?

Interesting with the attr spec'ing. Don't have time to test it atm, but I'll try to come back to this later, as I'm intrigued :)

**Comment by sczizzo** [comment]
Smort is one such application. Smort actually doesn't know about Chef at all, but we make API requests against Smort from Chef to pull in JSON documents. We might use those documents on their own or merge them into the node object. For the API request we have a simple function `smort_lookup` defined in a library:

```
module Smort
  require 'net/http'
  require 'json'

  def smort_lookup schema, document=node.environment, endpoint=node.smort.endpoint
    lookup = URI.join endpoint, ::File.join('l', schema, document)
    results = Net::HTTP.get lookup
    JSON.parse(results) rescue nil
  end
end
```

Then in a recipe we might use it like so:

```
metric_checks = smort_lookup 'metric_checks'

metric_checks.each do |check_defn|
  sensu_config check_defn['name'] do
    checks \
      check_defn['name'] => {
        type: 'check',
        standalone: true,
        handlers: [ check_defn['handler'] ],
        interval: check_defn['interval'],
        command: check_defn['command']
      }
  end
end
```

Chef is still writing our configuration (`sensu_config`), but the attributes are pulled from Smort rather than Chef Server.

**Comment by patcon** [comment]
Rockin! I don't want to presume you don't know about them, but here's some prior art just in case. Chef explicitly chose to bundle the config registry with the whole config management tool, but previous tools had separation, and some people have tried to split that back out with chef.

## Noah
http://blog.lusis.org/blog/2013/01/20/future-of-noah/ (good links at the bottom)
https://github.com/lusis/Noah
https://github.com/lusis-cookbooks/noah (the magic to replace attrs, iirc)

## etcd
If it seems like it doesn't version config, it just uses a wonky term called "index".
https://coreos.com/etcd/docs/2.0.8/api.html#setting-a-key-from-a-file

## zookeeper
Actually here's just a good SO question with links: https://stackoverflow.com/questions/6047917/zookeeper-alternatives-cluster-coordination-service
(Consul wasn't even in the picture when I was doing my own research, so that's how behind I am!)

Anyhow, curious to hear if I've misunderstood. The thought was perhaps it might make more sense to use these established tools and write a tool that handles the schema validation, but not the actual storage/notification. To clarify, this isn't a space I'm especially strong in, so I don't want to mislead! But the thought it that the core of this might be something that exists already for more general use at large scale.

**Comment by sczizzo** [comment]
Yeah I took a lot of inspiration from Consul's KV store (which looks like etcd's Index concept) when I wrote DBum (https://github.com/sczizzo/dbum). The original idea for Smort was it would just be an interface to define and manage schemas, and I would use DBum/Consul/&c as my database. Right now, Smort is implemented with a local DB, but I'm not doing anything too fancy; it should translate fairly easily to a different backing store.

**Comment by patcon** [comment]
Gotcha! Sorry, hard to know where everyone falls in terms of experience, so thought I'd be liberal with info. I'll follow along -- I'm sure I could learn something :)
